README.md

# laser.js

> laser-precision animation sequencing & timing

### Why?

**More Control:**

jQuery uses an animation queue by default. This is great if you're animating one or two items. If you have a very precise set of animations that you would like to create, modify, and call at an exact time in a sequence; this solves those problems.

### Tested Compatability:

  * IE8+
  * Chrome 27+
  * Safari 6+
  * Firefox 21+

### Usage ala [candidblend.github.io/laser.js](http://candidblend.github.io/laser.js)

```javascript

/**
 * @description basic example of laser.js capabilities/usage
 * @url http://candidblend.github.io/laser.js
 */

$(function() {

  // initialize new laser.js instance
  var seq = new Laser();

  // custom easing fn
  seq.addEasing('bounceOut', 'cubic-bezier(.33,1.66,.08,-1.71)');

  // on animation completed
  seq.on('animation:completed', function(id) {
    console.log('> animation completed with id: %s', id);
  });

  // on sequence start
  seq.on('sequence:started', function(total) {
    console.log('> sequence started with (%s) animations', total);
  });

  // on sequence complete
  seq.on('sequence:completed', function(elapsed) {
    console.log('> sequence completed in: %s milliseconds', elapsed);
  });

  // add to sequence, 'easing' defaults to 'easeOutExpo', 'when' to 0
  seq
    .add('#box-1', { top : 50,  left : 80 }, { duration : 1000 })
    .add('#box-2', { top : 100, left : 80 }, { duration : 1000 })
    .add('#box-3', { top : 150, left : 80 }, { duration : 1000 })
    .add('#box-4', { top : 200, left : 80 }, { duration : 1000 })
    .add('#box-5', { top : 250, left : 80 }, { duration : 1000 })
    .add('#box-6', { top : 300, left : 80 }, { duration : 1000 })

    .add('#box-1', { width : 40,  left : 0   }, { duration : 150,  when : 1250 })
    .add('#box-2', { width : 80,  left : 40  }, { duration : 300,  when : 1250 })
    .add('#box-3', { width : 120, left : 120 }, { duration : 450,  when : 1250 })
    .add('#box-4', { width : 160, left : 80  }, { duration : 600,  when : 1250 })
    .add('#box-5', { width : 200, left : 40  }, { duration : 750,  when : 1250 })
    .add('#box-6', { width : 240, left : 0   }, { duration : 1000, when : 1250 })

    .add('#box-1', { width : 40, left : 0   }, { duration : 1000, when : 2250 })
    .add('#box-2', { width : 40, left : 100 }, { duration : 750,  when : 2250 })
    .add('#box-3', { width : 40, left : 200 }, { duration : 600,  when : 2250 })
    .add('#box-4', { width : 40, left : 300 }, { duration : 450,  when : 2250 })
    .add('#box-5', { width : 40, left : 400 }, { duration : 300,  when : 2250 })
    .add('#box-6', { width : 40, left : 500 }, { duration : 150,  when : 2250 })

    .add('#box-1', { left : 275, top : 160 }, { duration : 250,  when : 3500 })
    .add('#box-2', { left : 275, top : 160 }, { duration : 500,  when : 3500 })
    .add('#box-3', { left : 275, top : 160 }, { duration : 750,  when : 3500 })
    .add('#box-4', { left : 275, top : 160 }, { duration : 1000, when : 3500 })
    .add('#box-5', { left : 275, top : 160 }, { duration : 1250, when : 3500 })
    .add('#box-6', { left : 275, top : 160 }, { duration : 1500, when : 3500 })

    .add('#box-1', { scale : 1.5 }, { duration : 250,  easing : 'bounceOut', when : 5000 })
    .add('#box-2', { scale : 3.0 }, { duration : 500,  easing : 'bounceOut', when : 5000 })
    .add('#box-3', { scale : 4.5 }, { duration : 750,  easing : 'bounceOut', when : 5000 })
    .add('#box-4', { scale : 6.0 }, { duration : 1000, easing : 'bounceOut', when : 5000 })
    .add('#box-5', { scale : 7.5 }, { duration : 1250, easing : 'bounceOut', when : 5000 })
    .add('#box-6', { scale : 8.0 }, { duration : 1500, easing : 'bounceOut', when : 5000 });

  // play sequence
  seq.play();

});

```

### Dependencies

**jQuery:**

Since `laser.js` is an extension of jQuery's .animate method, and is quite common in most projects, it's a dependency.

**Underscore.js:**

`underscore` until we don't need polyfills **:/**

### Highly Recommended:

**jQuery Transit**

[jQuery Transit](http://ricostacruz.com/jquery.transit/). **jQuery Transit** extends ***jQuery.animate*** with CSS3 transitions (hardware acceleration, speed, more types) where possible. Although not required, highly recommended.

### License (MIT)

> Copyright (c) 2013, Edward Hotchkiss



CHANGELOG.md

0.0.2 / 2013-07-06 
==================

 * [dist] build bump
 * [bower] match bower.json to --tags
 * [bower] json parse
 * [changelog] updated
 * [dist] bump to 0.0.2
 * [bower, polyfills, test] added basic tests, simple test suite, using bower now
 * [demo] move to gh-pages and edit docs to reflect new demo src

0.0.1 / 2013-06-27 
==================

 * [go time]


test/js/spec/basic.js

/**
 * @description basic test coverage
 */

$(function() {

  new Kale('basic laser.js DOM related sync & async tests')

    .setup([ /* ----- */ ], function() {
      var $box = $('<div id="box"></div>')
        .css({
          top        : '50px',
          left       : '50px',
          width      : '50px',
          height     : '50px',
          border     : '1px solid #dddddd',
          position   : 'absolute',
          background : '#eeeeee'
        })
        .appendTo($('body'));
    })

    .addTest('#box should be present in the DOM', function() {
      this.assertInDOM('#box');
    })

    .addTest('#box should be present in the DOM, even after 100ms', function() {
      var _this = this;
      setTimeout(function() {
        _this.assertInDOM('#box');
      }, 100);
    })

    .addTest('#boxx should NOT be present in the DOM', function() {
      this.assertNotInDOM('#boxx');
    })

    .addTest('#boxx should be present in the DOM, even after 100ms', function() {
      var _this = this;
      setTimeout(function() {
        _this.assertNotInDOM('#boxx');
      }, 100);
    })

    .addTest('laser should be able to create new instances', function() {
      this.assertTrue(new Laser() instanceof Laser);
    })

    .addTest('#box should be animated to `left` `500px` after 300ms', function() {
      var basicSeq = new Laser();
      basicSeq.add('#box', { left : 500 }, { duration : 300, when : 0 });
      basicSeq.on('sequence:completed', _.bind(function() {
        this.assertEqual($('#box').css('left'), '500px');
      }, this));
      basicSeq.play();
    })

    // `true` flag means remove all .setup resources / reset the DOM to an emtpty state
    .tearDown(true, function() {
      /* ----- */
    })

    .run();

});

/* EOF */

Gruntfile.js

/**
 * @description Watch, concat and minify for dist.
 */

module.exports = function(grunt) {

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    uglify: {
      options: {
        banner: '/*! laser.js v<%= pkg.version %>, <%= grunt.template.today("yyyy-mm-dd") %> */\n'
      },
      build: {
        src: 'dist/laser-<%= pkg.version %>.js',
        dest: 'dist/laser-<%= pkg.version %>.min.js'
      }
    },
    jshint: {
      options: {
        browser: true,
        globals: {
          jQuery: true
        }
      },
      files: {
        src: [
          'Gruntfile.js',
          'src/laser.js',
          'test/js/vendor/kale.js',
          'test/js/spec/basic.js'
        ]
      }
    },
    concat: {
      options: {
      },
      dist: {
        src: ['src/laser.js'],
        dest: 'dist/laser-<%= pkg.version %>.js'
      }
    },
    watch: {
      scripts: {
        files: [
          'Gruntfile.js',
          'src/laser.js',
          'test/js/vendor/kale.js',
          'test/js/spec/basic.js'
        ],
        tasks: ['jshint','concat']
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-watch');

  grunt.registerTask('build', ['jshint','concat','uglify']);
  grunt.registerTask('default', ['jshint','watch']);

};

/* EOF */

src/laser.js

/**
 * @library laser.js
 * @author Edward Hotchkiss <edward@candidblend.la>
 * @description Laser-precision animation sequencing and chaining
 * with granular controls on a strong Event based architecture.
 * @license MIT
 */

(function() {

  'use strict';

  /**
   * @method Laser
   * @description constructor fn
   */

  var Laser = window.Laser = function Laser(params) {
    this.cache = {};
    this.listeners = {};
    this.animations = [];
    return _.extend(this, params);
  };

  /**
   * @private _isTransition
   * @description determine whether jQuery transit is available
   * @return {Boolean} availability
   */

  function _isTransition(alias, easing) {
    return (typeof(jQuery().transition) === 'function') ? true : false;
  }

  /**
   * @private _isValidEasing
   * @description invalid easing method aliases bork .animate/.transition
   * check that the alias exists in the dictionary
   * @param {String} ea easing alias
   * @return {Boolean} exists or doesn't
   */

  function _isValidEasing(alias) {
    if (_isTransition() === true) {
      // css3 easing cubic-bezier strings
      return ($.cssEase[alias] !== undefined) ? true : false;
    } else if (_isTransition() === false) {
      // standard easing fns
      return ($.easing[alias] !== undefined) ? true : false;
    }
  }

  /**
   * @description extend our constructor protoype with public methods
   */

  _.extend(Laser.prototype, {

    /**
     * @method get
     * @param {String} attr instance attribute
     * @param {Object} where set of key/values to match
     * @description instance attribute getter
     */

    get: function(attr, where) {
      if (where === undefined) {
        return this[attr];
      } else {
        return _.where(this[attr], where);
      }
    },

    /**
     * @method set
     * @param {String} attr instance attribute
     * @param {Object} where set of key/values to match
     * @param {Object} params set of key/values to set
     * @description instance attribute getter
     */

    set: function(attr, where, params) {
      var item = this.get(attr, where);
      if (item === undefined) {
        return undefined;
      } else {
        _.forEach(params, function(val, key) {
          result[key] = val;
        }, ctx);
        return obj;
      }
    },

    /**
     * @method getCache
     * @param {String} s css selector
     * @description gets cached jQuery element
     * @return {Function} jQuery wrapped element
     */

    getCache: function(s) {
      return this.cache[s];
    },

    /**
     * @method setCache
     * @param {String} s css selector
     * @description first checks for cached jQuery element by selector,
     * otherwise caches reference to the jQuery element
     * @return {Function} jQuery wrapped element
     */

    setCache: function(s) {
      this.cache[s] = this.getCache(s) || $(s);
      return this.cache[s];
    },

    /**
     * @method on
     * @param {String} n event name
     * @param {Function} fn trigger function to store
     * @description bind function to event name
     */

    on: function(n, fn) {
      this.listeners[n] = this.listeners[n] || [];
      this.listeners[n].push(fn);
    },

    /**
     * @method off
     * @param {String} n event name
     * @param {Function} fn trigger function to remove
     * @description remove event listener
     */

    off: function(n, fn) {
      if (this.listeners[n]) {
        this.listeners[n].splice(this.listeners[n].indexOf(fn), 1);
      }
    },

    /**
     * @method trigger
     * @param {String} n event name
     * @description trigger event listener
     */

    trigger: function(n) {
      if (this.listeners[n]) {
        var args = Array.prototype.slice.call(arguments, 1);
        _.forEach(this.listeners[n], function(v, i, obj) {
          v.apply(this, args);
        }, this);
      }
    },

    /**
     * @method id
     * @description generate a new id. method exists
     * in the case they we switch to a UUID in the future
     * @return {Number} animation's index
     */

    id: function() {
      return this.get('animations').length;
    },

    /**
     * @method _createAnimation
     * @description creates a new animation object
     * @param {Object} obj params
     * @return {Object} jQuery ready animation obj
     */

    createAnimation: function(obj) {
      // complete callback
      obj.options.complete = _.bind(function() {
        obj.state = 'COMPLETE';
        this.remaining--;
        if (this.remaining === 0) {
          this.finishedAt = new Date().getTime();
          var elapsed = this.finishedAt - this.startedAt;
          this.trigger('sequence:completed', elapsed);
        }
        this.trigger('animation:completed', obj.id);
        clearTimeout(obj.timeout);
      }, this);
      // fn call entry point
      obj.fn = _.bind(function() {
        // timeout reference for future 'when'
        obj.timeout = setTimeout(_.bind(function() {
          if (_isTransition() === true) {
            obj.$elem.transition(obj.params, obj.options.duration, obj.options.easing, obj.options.complete);
          } else {
            obj.$elem.animate(obj.params, obj.options);
          }
        }, this), obj.when);
        obj.state = 'ON_STACK';
      }, this);
      return obj;
    },

    /**
     * @method addEasing
     * @description add either a css3 cubic-bezier ease or a jquery easing fn
     */

    addEasing: function(alias, easing) {
      if (typeof(easing) === 'string') {
        $.cssEase[alias] = easing;
      } else {
        $.easing[alias] = easing;
      }
      return this;
    },

    /**
     * @method add
     * @description sets up params via arguments for a
     * new animation object to push onto the sequence stack
     * @param {String} s css selector for element to animate
     * @param {Object} p jQuery standard animation parameters
     * @param {Object} o jQuery standard animation options,
     * excluding the 'when' attribute
     */

    add: function(s, p, o) {
      var w, i = this.id(), e = this.setCache(s);
      // get then remove 'when' from standard options
      w = o.when || 0;
      delete o.when;
      // defaults
      o.queue = false;
      o.easing = o.easing || 'easeOutExpo';
      if (!_isValidEasing(o.easing)) {
        throw new Error('Easing method not defined! - '+o.easing);
      }
      o.duration = o.duration || 500;
      // push new animation onto sequence
      this.animations.push(
        this.createAnimation({
          id       : i,
          when     : w,
          params   : p,
          options  : o,
          $elem    : e
        })
      );
      return this;
    },

    /**
     * @method play
     * @description plays animation sequence
     */
  
    play: function() {
      var stack = this.get('animations');
      this.startedAt = new Date().getTime();
      this.remaining = stack.length;
      this.trigger('sequence:started', stack.length);
      _.map(stack, function(val, index) {
        stack[index].fn.call(this);
      }, this);
      return this;
    },

    /**
     * @method stop
     * @param {Number} when time in milliseconds through sequence to stop at
     * @description clear all active animations, and stop currents
     */

    stop: function(when) {
      var timeout = setTimeout(_.bind(function() {
        clearTimeout(timeout);
        _.map(this.get('animations', { state : 'ON_STACK' }), function(item) {
          clearTimeout(item.timeout);
          item.$elem.stop(true, false);
          item.state = 'STOPPED';
          this.trigger('animation:stopped', item.id);
        }, this);
      }, this), when || 0);
      return this;
    }

  });
  
}());

/* EOF */

demo/js/src/basic.js

/**
 * @description basic example of laser.js capabilities/usage
 */

$(function() {

  // initialize new Laser instance
  var boxDemoSequence = new Laser();
  
  // animation completed
  boxDemoSequence.on('animation:completed', function(id) {
    console.log('> animation completed with id: %s', id);
  });

  // sequence started playing
  boxDemoSequence.on('sequence:started', function(count) {
    console.log('> sequence started with (%s) animations', count);
  });

  // sequence completed
  boxDemoSequence.on('sequence:completed', function(elapsed) {
    console.log('> sequence completed in: %s milliseconds', elapsed);
  });
  
  // the 'when' param defaults to 0
  boxDemoSequence
    .add('#box-1', { left : 100, top : 160 }, { duration : 250 })
    .add('#box-2', { left : 280, top : 160 }, { duration : 750 })
    .add('#box-3', { left : 460, top : 160 }, { duration : 500 })

    .add('#box-1', { scale : 1.5 }, { duration : 500, easing : 'easeOutBack', when : 750 })
    .add('#box-2', { scale : 2.0 }, { duration : 500, easing : 'easeOutBack', when : 750 })
    .add('#box-3', { scale : 2.5 }, { duration : 500, easing : 'easeOutBack', when : 750 })

    .add('#box-1', { scale : 1 }, { duration : 500, easing : 'easeInBack', when : 1250 })
    .add('#box-2', { scale : 1 }, { duration : 500, easing : 'easeInBack', when : 1250 })
    .add('#box-3', { scale : 1 }, { duration : 500, easing : 'easeInBack', when : 1250 })

    .add('#box-1', { left : 0, width : 30, height : 30 }, { duration : 1000, when : 2500 })
    .add('#box-2', { left : 40, width : 60, height : 60 }, { duration : 1000, when : 2500 })
    .add('#box-3', { left : 110, width : 90, height : 90 }, { duration : 1000, when : 2500 })

    .add('#box-1', { top : 0, width : 40, height : 40 }, { duration : 500, when : 3700 })
    .add('#box-2', { top : 0, left : 50, width : 40, height : 40 }, { duration : 500, when : 3700 })
    .add('#box-3', { top : 0, left : 100, width : 40, height : 40 }, { duration : 500, when : 3700 })

    .add('#box-1', { opacity : 0.5 }, { duration : 100, when : 4500 })
    .add('#box-2', { opacity : 0.5 }, { duration : 100, when : 4700 })
    .add('#box-3', { opacity : 0.5 }, { duration : 100, when : 4900 })

    .add('#box-1', { opacity : 1 }, { duration : 100, when : 5100 })
    .add('#box-2', { opacity : 1 }, { duration : 100, when : 5300 })
    .add('#box-3', { opacity : 1 }, { duration : 100, when : 5500 });

    // play sequence
    boxDemoSequence.play();

});

/* EOF */

js/src/page.js

/**
 * @description basic example of laser.js capabilities/usage
 */

$(function() {

  // initialize new laser.js instance
  var seq = new Laser();

  // custom easing fn
  seq.addEasing('bounceOut', 'cubic-bezier(.33,1.66,.08,-1.71)');

  // on animation completed
  seq.on('animation:completed', function(id) {
    console.log('> animation completed with id: %s', id);
  });

  // on sequence start
  seq.on('sequence:started', function(total) {
    console.log('> sequence started with (%s) animations', total);
  });

  // on sequence complete
  seq.on('sequence:completed', function(elapsed) {
    console.log('> sequence completed in: %s milliseconds', elapsed);
  });

  // add to sequence, easing defaults to 'easeOutExpo'
  seq
    .add('#box-1', { top : 50,  left : 80 }, { duration : 1000, when : 0 })
    .add('#box-2', { top : 100, left : 80 }, { duration : 1000, when : 0 })
    .add('#box-3', { top : 150, left : 80 }, { duration : 1000, when : 0 })
    .add('#box-4', { top : 200, left : 80 }, { duration : 1000, when : 0 })
    .add('#box-5', { top : 250, left : 80 }, { duration : 1000, when : 0 })
    .add('#box-6', { top : 300, left : 80 }, { duration : 1000, when : 0 })

    .add('#box-1', { width : 40,  left : 0   }, { duration : 150,  when : 1250 })
    .add('#box-2', { width : 80,  left : 40  }, { duration : 300,  when : 1250 })
    .add('#box-3', { width : 120, left : 120 }, { duration : 450,  when : 1250 })
    .add('#box-4', { width : 160, left : 80  }, { duration : 600,  when : 1250 })
    .add('#box-5', { width : 200, left : 40  }, { duration : 750,  when : 1250 })
    .add('#box-6', { width : 240, left : 0   }, { duration : 1000, when : 1250 })

    .add('#box-1', { width : 40, left : 0   }, { duration : 1000, when : 2250 })
    .add('#box-2', { width : 40, left : 100 }, { duration : 750,  when : 2250 })
    .add('#box-3', { width : 40, left : 200 }, { duration : 600,  when : 2250 })
    .add('#box-4', { width : 40, left : 300 }, { duration : 450,  when : 2250 })
    .add('#box-5', { width : 40, left : 400 }, { duration : 300,  when : 2250 })
    .add('#box-6', { width : 40, left : 500 }, { duration : 150,  when : 2250 })

    .add('#box-1', { left : 275, top : 160 }, { duration : 250,  when : 3500 })
    .add('#box-2', { left : 275, top : 160 }, { duration : 500,  when : 3500 })
    .add('#box-3', { left : 275, top : 160 }, { duration : 750,  when : 3500 })
    .add('#box-4', { left : 275, top : 160 }, { duration : 1000, when : 3500 })
    .add('#box-5', { left : 275, top : 160 }, { duration : 1250, when : 3500 })
    .add('#box-6', { left : 275, top : 160 }, { duration : 1500, when : 3500 })

    .add('#box-1', { scale : 1.5 }, { duration : 250,  easing : 'bounceOut', when : 5000 })
    .add('#box-2', { scale : 3.0 }, { duration : 500,  easing : 'bounceOut', when : 5000 })
    .add('#box-3', { scale : 4.5 }, { duration : 750,  easing : 'bounceOut', when : 5000 })
    .add('#box-4', { scale : 6.0 }, { duration : 1000, easing : 'bounceOut', when : 5000 })
    .add('#box-5', { scale : 7.5 }, { duration : 1250, easing : 'bounceOut', when : 5000 })
    .add('#box-6', { scale : 8.0 }, { duration : 1500, easing : 'bounceOut', when : 5000 });

  // play sequence
  seq.play();

});

/* EOF */